pkg/autodiscovery/autoconfig.go

```diff
@@ -102,27 +111,51 @@ func (ac *AutoConfig) StartPolling() {
 	ac.pollerActive = true
 }
 
+// startServiceListening waits on services and templates and process them as they come.
+// It can trigger scheduling decisions or just update its cache.
+func (ac *AutoConfig) startServiceListening() {
+	ac.m.Lock()
+	defer ac.m.Unlock()
+
+	go func() {
+		for {
+			select {
+			case <-ac.listenerStop:
+				ac.healthListening.Deregister()
+				return
+			case <-ac.healthListening.C:
+			case svc := <-ac.newService:
+				ac.processNewService(svc)
+			case svc := <-ac.delService:
+				ac.processDelService(svc)
+			}
+		}
+	}()
+	ac.listenerActive = true
+}
+
 // Stop just shuts down AutoConfig in a clean way.
 // AutoConfig is not supposed to be restarted, so this is expected
 // to be called only once at program exit.
 func (ac *AutoConfig) Stop() {
 	ac.m.Lock()
 	defer ac.m.Unlock()
 
-	// stop the poller if running
+	// stop the config poller if running
 	if ac.pollerActive {
 		ac.pollerStop <- struct{}{}
 		ac.pollerActive = false
 	}
 
+	// stop the service listener if running
+	if ac.listenerActive {
```

Comment:
I tend to avoid using this kind of check, why could the listener be inactive here ? Why would we stop a component that we did not start ahead ?

Reply:
Yeah initially, I was not starting the service listening right away so if you stopped the agent in the meanwhile, this could happen.
Now it's starting as we build the object so I'll completely drop the `bool`

Reply:
👍